### PR TITLE
Add support for MutatingWebhookConfiguration +ValidatingWebhookConfiguration

### PR DESCRIPTION
**Pull request by natalysheinin**
**What this PR does / why we need it**: Feature, adds support for generating metrics for existing kubernetes resources: `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration`

https://godoc.org/k8s.io/api/admissionregistration/v1beta1#MutatingWebhookConfiguration
https://godoc.org/k8s.io/api/admissionregistration/v1beta1#ValidatingWebhookConfiguration

**Which issue(s) this PR fixes**
Fixes https://github.com/kubernetes/kube-state-metrics/issues/853

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,27 +2,20 @@ module k8s.io/kube-state-metrics
...
-	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
-	k8s.io/api v0.0.0-20190620084959-7cf5895f2711
-	k8s.io/apimachinery v0.0.0-20190612205821-1799e75a0719
+	k8s.io/api v0.0.0-20190927115716-5d581ce610b0
+	k8s.io/apimachinery v0.0.0-20190927035529-0104e33c351d
 	k8s.io/autoscaler v0.0.0-20190607113959-1b4f1855cb8e
 	k8s.io/client-go v0.0.0-20190620085101-78d2af792bab
-	k8s.io/klog v0.4.0
-	k8s.io/kube-openapi v0.0.0-20190320154901-5e45bb682580 // indirect
+	k8s.io/klog v1.0.0
```
Comment (tariq1890) [inline go.mod]: I don't think we should be changing the `go.mod` file for this. Can we work on top of the existing deps for this PR?

File: docs/validatingwebhookconfiguration.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# ValidatingWebhookConfiguration Metrics
+
+| Metric name| Metric type | Labels/tags | Status |
+| ---------- | ----------- | ----------- | ----------- |
+| kube_validatingwebhookconfiguration_info | Gauge | `validatingwebhookconfiguration`=&lt;validatingwebhookconfiguration-name&gt; <br> `namespace`=&lt;validatingwebhookconfiguration-namespace&gt; | STABLE |
```
Comment (tariq1890) [inline docs/validatingwebhookconfiguration.md]: New metrics have to be `EXPERIMENTAL`. Apologies for not being clear. We are yet to add documentation to outline this.
```suggestion
| kube_validatingwebhookconfiguration_info | Gauge | `validatingwebhookconfiguration`=&lt;validatingwebhookconfiguration-name&gt; <br> `namespace`=&lt;validatingwebhookconfiguration-namespace&gt; | EXPERIMENTAL |
```

File: docs/mutatingwebhookconfiguration.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# MutatingWebhookConfiguration Metrics
+
+| Metric name| Metric type | Labels/tags | Status |
+| ---------- | ----------- | ----------- | ----------- |
+| kube_mutatingwebhookconfiguration_info | Gauge | `mutatingwebhookconfiguration`=&lt;mutatingwebhookconfiguration-name&gt; <br> `namespace`=&lt;mutatingwebhookconfiguration-namespace&gt; | STABLE |
```
Comment (tariq1890) [inline docs/mutatingwebhookconfiguration.md]: New metrics have to be EXPERIMENTAL. Apologies for not being clear. We are yet to add documentation to outline this.
```suggestion
| kube_mutatingwebhookconfiguration_info | Gauge | `mutatingwebhookconfiguration`=&lt;mutatingwebhookconfiguration-name&gt; <br> `namespace`=&lt;mutatingwebhookconfiguration-namespace&gt; | EXPERIMENTAL |
```

File: internal/store/mutatingwebhookconfiguration.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+var (
+	descMutatingWebhookConfigurationName          = "kube_mutatingwebhookconfiguration_labels"
+	descMutatingWebhookConfigurationHelp          = "Kubernetes labels converted to Prometheus labels."
+	descMutatingWebhookConfigurationDefaultLabels = []string{"namespace", "mutatingwebhookconfiguration"}
+
+	mutatingwebhookconfigurationFamilies = []metric.FamilyGenerator{
+		{
+			Name: "kube_mutatingwebhookconfiguration_info",
+			Type: metric.Gauge,
+			Help: "Information about mutatingwebhookconfiguration.",
```
Comment (tariq1890) [inline internal/store/mutatingwebhookconfiguration.go]:
```suggestion
			Help: "Information about the Mutating Webhook Configuration.",
```

File: internal/store/mutatingwebhookconfiguration.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+package store
+
+import (
+	admissionregistration "k8s.io/api/admissionregistration/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kube-state-metrics/pkg/metric"
```
Comment (tariq1890) [inline internal/store/mutatingwebhookconfiguration.go]:
```suggestion

	"k8s.io/kube-state-metrics/pkg/metric"
```

File: internal/store/mutatingwebhookconfiguration_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+*/
+
+package store
+
+import (
+	"testing"
+
+	admissionregistration "k8s.io/api/admissionregistration/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kube-state-metrics/pkg/metric"
```
Comment (tariq1890) [inline internal/store/mutatingwebhookconfiguration_test.go]:
```suggestion

	"k8s.io/kube-state-metrics/pkg/metric"
```

File: internal/store/validatingwebhookconfiguration.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+package store
+
+import (
+	admissionregistration "k8s.io/api/admissionregistration/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kube-state-metrics/pkg/metric"
```
Comment (tariq1890) [inline internal/store/validatingwebhookconfiguration.go]:
```suggestion

	"k8s.io/kube-state-metrics/pkg/metric"
```

File: internal/store/validatingwebhookconfiguration.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+var (
+	descValidatingWebhookConfigurationName          = "kube_validatingwebhookconfiguration_labels"
+	descValidatingWebhookConfigurationHelp          = "Kubernetes labels converted to Prometheus labels."
+	descValidatingWebhookConfigurationDefaultLabels = []string{"namespace", "validatingwebhookconfiguration"}
+
+	validatingwebhookconfigurationFamilies = []metric.FamilyGenerator{
+		{
+			Name: "kube_validatingwebhookconfiguration_info",
+			Type: metric.Gauge,
+			Help: "Information about the ValidatingWebhookConfiguration.",
```
Comment (tariq1890) [inline internal/store/validatingwebhookconfiguration.go]:
```suggestion
			Help: "Information about the validating webhook configuration.",
```

File: internal/store/validatingwebhookconfiguration.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_validatingwebhookconfiguration_metadata_resource_version",
+			Type: metric.Gauge,
+			Help: "Resource version representing a specific version of the ValidatingWebhookConfiguration.",
```
Comment (tariq1890) [inline internal/store/validatingwebhookconfiguration.go]:
```suggestion
			Help: "Resource version representing a specific version of the validating webhook configuration.",
```

File: internal/store/mutatingwebhookconfiguration.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+var (
+	descMutatingWebhookConfigurationName          = "kube_mutatingwebhookconfiguration_labels"
+	descMutatingWebhookConfigurationHelp          = "Kubernetes labels converted to Prometheus labels."
+	descMutatingWebhookConfigurationDefaultLabels = []string{"namespace", "mutatingwebhookconfiguration"}
+
+	mutatingwebhookconfigurationFamilies = []metric.FamilyGenerator{
+		{
+			Name: "kube_mutatingwebhookconfiguration_info",
+			Type: metric.Gauge,
+			Help: "Information about the MutatingWebhookConfiguration.",
```
Comment (tariq1890) [inline internal/store/mutatingwebhookconfiguration.go]:
```suggestion
			Help: "Information about the mutating webhook configuration.",
```

File: internal/store/mutatingwebhookconfiguration.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_mutatingwebhookconfiguration_metadata_resource_version",
+			Type: metric.Gauge,
+			Help: "Resource version representing a specific version of the MutatingWebhookConfiguration.",
```
Comment (tariq1890) [inline internal/store/mutatingwebhookconfiguration.go]:
```suggestion
			Help: "Resource version representing a specific version of the mutating webhook configuration.",
```

**Comment by tariq1890** [comment]
@natalysheinin You will also need to update the ClusterRole to add list/watch permissions on Validating and Mutating Webhook configurations.

File: internal/store/validatingwebhookconfiguration.go (hunk 9 above)
Comment (natalysheinin) [reply internal/store/validatingwebhookconfiguration.go]: Why space separate it? The official resource name is `MutatingWebhookConfiguration` in the k8s API. 

Also, keeping it as one word will keep it consistent with other resource representations in this repo. 
[example](https://github.com/kubernetes/kube-state-metrics/blob/master/internal/store/persistentvolume.go#L102)

Comment (tariq1890) [reply internal/store/validatingwebhookconfiguration.go]: Looks like we haven't been too consistent with the naming. There are other examples where we space-separate it, for e.g:- limit ranges, resource quota, persistent volume claims. So either as-is/space-separating would work.

File: examples/standard/cluster-role.yaml [hunk 12]
```diff
@@ -92,3 +92,16 @@ rules:
...
   - watch
+- apiGroups:
+  - admissionregistration.k8s.io
+  resources:
+  - mutatingwebhookconfigurations
+  - validatingwebhookconfigurations
+  verbs:
+  - list
+  - watch
+
```
Comment (tariq1890) [inline examples/standard/cluster-role.yaml]: These empty lines are extraneous. We should have only one empty line for the EOF.

File: internal/store/mutatingwebhookconfiguration.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+	mutatingwebhookconfigurationFamilies = []metric.FamilyGenerator{
+		{
+			Name: "kube_mutatingwebhookconfiguration_info",
+			Type: metric.Gauge,
+			Help: "Information about the MutatingWebhookConfiguration.",
+			GenerateFunc: wrapMWCFunc(func(mwc *admissionregistration.MutatingWebhookConfiguration) *metric.Family {
+				return &metric.Family{
+					Metrics: []*metric.Metric{{
+						LabelKeys:   []string{},
```
Comment (tariq1890) [inline internal/store/mutatingwebhookconfiguration.go]: We don't need these. Let's remove it here as well as in `validatewebhookconfiguration` as it's implied. 
```
LabelKeys:   []string{},
LabelValues: []string{},
```

File: examples/standard/cluster-role.yaml (hunk 12 above)
Comment (natalysheinin) [reply examples/standard/cluster-role.yaml]: I believe there is only one empty line atm

File: docs/README.md [hunk 14]
```diff
@@ -45,11 +45,18 @@ Stages about metrics are grouped into three categories：
 
 Per group of metrics there is one file for each metrics. See each file for specific documentation about the exposed metrics:
 
+- [CertificateSigningRequest Metrics](certificatessigningrequest-metrics.md)
```
Comment (natalysheinin) [inline docs/README.md]: re-organized so they are actually in alphabetical order

File: internal/store/mutatingwebhookconfiguration.go (hunk 13 above)
Comment (natalysheinin) [reply internal/store/mutatingwebhookconfiguration.go]: thanks

**Comment by natalysheinin** [comment]
@tariq1890 not sure why lint is upset, 
```
internal/store/mutatingwebhookconfiguration.go:42: File is not `goimports`-ed (goimports)
						Value:       1,
```

do you know what I need to do?

File: docs/README.md (hunk 14 above)
Comment (tariq1890) [reply docs/README.md]: Yay 🎉 Thank you :)

File: examples/standard/cluster-role.yaml (hunk 12 above)
Comment (tariq1890) [reply examples/standard/cluster-role.yaml]: I see more in my diff. L104-107.  Is that not the case for you?

Comment (natalysheinin) [reply examples/standard/cluster-role.yaml]: aah yes. I was looking at the other cluster-role file 🤦‍♀️

**Comment by tariq1890** [comment]
@natalysheinin Looks like you caught the issue!

**Comment by natalysheinin** [comment]
Everything but e2e tests are passing: https://travis-ci.org/kubernetes/kube-state-metrics/jobs/592765966
```
W1002 20:55:39.619189    4019 root.go:254] Error reading config file at /home/travis/.minikube/config/config.json: open /home/travis/.minikube/config/config.json: no such file or directory
```

not sure this is related to me.

**Comment by tariq1890** [comment]
@natalysheinin Forgot to mention - We need to add your new k8s resources [here](https://github.com/kubernetes/kube-state-metrics/blob/master/pkg/options/collector.go#L37).

**Comment by tariq1890** [comment]
> not sure this is related to me.

Yup, that's unrelated.

**Comment by tariq1890** [comment]
@natalysheinin We need to add the builder methods as well. Sorry, I forgot to mention that. You can see over [here](https://github.com/kubernetes/kube-state-metrics/blob/d10e79428289c364936d6e66e8359e8d194bea3d/internal/store/builder.go#L276).

File: internal/store/builder.go [hunk 15]
```diff
@@ -221,6 +223,10 @@ func (b *Builder) buildLimitRangeStore() *metricsstore.MetricsStore {
 	return b.buildStore(limitRangeMetricFamilies, &v1.LimitRange{}, createLimitRangeListWatch)
 }
 
+func (b *Builder) buildMWCStore() *metricsstore.MetricsStore {
+	return b.buildStore(namespaceMetricFamilies, &v1.Namespace{}, createNamespaceListWatch)
```
Comment (tariq1890) [inline internal/store/builder.go]: I'd prefer clarity to brevity here. Let's name it something like `buildMutatingWebhookConfigurationStore` or `buildMutatingWebhookConfigStore`.

Also, you should be creating a ListWatch method specific to your resource. Here you have re-used the list watch method of namespace.

File: internal/store/builder.go [hunk 16]
```diff
@@ -277,6 +283,10 @@ func (b *Builder) buildCsrStore() *metricsstore.MetricsStore {
 	return b.buildStore(csrMetricFamilies, &certv1beta1.CertificateSigningRequest{}, createCSRListWatch)
 }
 
+func (b *Builder) buildVWCStore() *metricsstore.MetricsStore {
```
Comment (tariq1890) [inline internal/store/builder.go]: Similar comment as above.

**Comment by tariq1890** [comment]
Once the code passes  the e2e tests, please don't forget to squash your commits.

**Comment by natalysheinin** [comment]
@tariq1890 anything I can do to fix the e2e tests?

```
W1003 14:32:09.720989    4009 root.go:254] Error reading config file at /home/travis/.minikube/config/config.json: open /home/travis/.minikube/config/config.json: no such file or directory
```

**Comment by tariq1890** [comment]
@natalysheinin The E2e tests are failing since it's looking for MutatingWebhook and ValidatingWebhook metrics. If you create sample manifests of the two resources inside this [directory](https://github.com/kubernetes/kube-state-metrics/tree/master/tests/manifests), the e2e tests will capture the new metrics and pass.

**Comment by natalysheinin** [comment]
@tariq1890 I am ready to squash and merge once you approve

**Comment by natalysheinin** [comment]
Thanks for all your help @tariq1890 , it seems like the CI requires an approved label and apparently I don't have permissions to merge it either.

**Comment by tariq1890** [comment]
Yup, I'm aware. I was actually waiting for you to squash the commits :).

**Comment by tariq1890** [comment]
Perfect, Thanks again!

/lgtm
/approve

**Comment by natalysheinin** [comment]
@tariq1890 do you know which version release will this feature be included in?

**Comment by LiliC** [comment]
@natalysheinin @tariq1890  We should be including this in the next release, lets do one soon as we have a lot of changes since the last one. WDYT?

**Comment by tariq1890** [comment]
@LiliC Sure we could cut the next release. I was thinking of waiting until K8s 1.16 was vendored in. But I don’t feel too strongly about it.

@brancz thoughts?

**Comment by brancz** [comment]
No strong opinion, although I don't feel that much has happened since the last release, so I'd prefer to wait for a couple more things to land.

**Pull request merged**